### PR TITLE
mypy: array_api

### PR DESCRIPTION
**Pull request by asi1024**
Depends on #6486.

### CODE REVIEW

**Comment by kmaehashi** [comment]
Should we upstream this instead of fixing in CuPy?

**Review by toslunar — APPROVED** [review]
LGTM

**Comment by leofang** [comment]
cc: @asmeurer @BvB93

File: cupy/array_api/_data_type_functions.py [hunk 1]
```diff
@@ -96,7 +94,7 @@ def finfo(type: Union[Dtype, Array], /) -> finfo_object:
 
     See its docstring for more information.
     """
-    fi = np.finfo(type)
+    fi = np.finfo(type)  # type: ignore
```
Comment (BvB93) [inline cupy/array_api/_data_type_functions.py:97]: The errors for `finfo` and `iinfo` are more legit though. The issue stems from the `Array.dtype` property being annotated as [returning a union](https://github.com/cupy/cupy/blob/ec6d2c9a98440b19a133868576fb299972c21f6c/cupy/array_api/_typing.py#L56-L67), which _always_ includes both floating- and integer-types while the `iinfo`/`finfo` constructors only accept one category or the other. 
This is a demonstration why it's usually a bad idea to return these so-called "unsafe" unions, and I would recommend to simply switch the dtype parameter over to `Any`:

``` diff
diff --git a/cupy/array_api/_array_object.py b/cupy/array_api/_array_object.py
index c2ae0545d..d713e927a 100644
--- a/cupy/array_api/_array_object.py
+++ b/cupy/array_api/_array_object.py
@@ -1047,7 +1047,7 @@ class Array:
             return Array._new(arr)

     @property
-    def dtype(self) -> Dtype:
+    def dtype(self) -> np.dtype[Any]:
         """
         Array API compatible wrapper for :py:meth:`np.ndarray.dtype <numpy.ndarray.dtype>`.
```

File: cupy/array_api/_array_object.py [hunk 2]
```diff
@@ -442,7 +441,7 @@ def __dlpack_device__(self: Array, /) -> Tuple[IntEnum, int]:
         """
         return self._array.__dlpack_device__()
 
-    def __eq__(self: Array, other: Union[int, float, bool, Array], /) -> Array:
+    def __eq__(self: Array, other: Union[int, float, bool, Array], /) -> Array:  # type: ignore
```
Comment (BvB93) [inline cupy/array_api/_array_object.py:444]: Yeah, this is an unfortunate side-effect of being incompatible with the (somewhat overly strict) [`object.__eq__`/`__ne__`](https://github.com/python/typeshed/blob/9687d53b65d055a990f0187b7baa1137cb7c717b/stdlib/builtins.pyi#L102-L103) annotations. Nothing to do about this though except adding a `# type: ignore` comment or complain a bit more upstream...

**Review by BvB93 — COMMENTED** [review]
Thanks for the ping @leofang!

**Pull request merged**